tools/x-repo-types/xrt.go

```diff
@@ -0,0 +1,244 @@
+	rootCmd.Flags().StringVar(&xBranch, "x-branch", "", "repository branch for type x")
+	rootCmd.Flags().StringVar(&xType, "x-type", "", "Exported type in the package for type x")
+	rootCmd.Flags().StringVar(&yPkg, "y-package", "", "Go repo and package for type for type y")
+	rootCmd.Flags().StringVar(&yBranch, "y-branch", "", "repository branch for type y")
+	rootCmd.Flags().StringVar(&yType, "y-type", "", "Exported type in the package for type y")
+
+	if err := rootCmd.Execute(); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func runApp(xPkg, xBranch, xType, yPkg, yBranch, yType string) error {
+	if xPkg == "" || xType == "" {
+		return fmt.Errorf("package:%s, and type:%s flags are required", xPkg, xType)
+	}
+	if yPkg == "" || yType == "" {
+		return fmt.Errorf("package:%s, and type:%s flags are required", yPkg, yType)
+	}
+
+	xPkgBranch := xPkg
+	if xBranch != "" {
+		xPkgBranch += "@" + xBranch
+	}
+	yPkgBranch := yPkg
+	if yBranch != "" {
+		yPkgBranch += "@" + yBranch
+	}
+
+	err := goGet(xPkgBranch)
+	if err != nil {
+		return err
+	}
+	err = goGet(yPkgBranch)
+	if err != nil {
+		return err
+	}
+
+	// Show the type/outline
+	err = showKind(xPkg, xType)
+	if err != nil {
+		return err
+	}
+	err = showKind(yPkg, yType)
+	if err != nil {
+		return err
+	}
+
+	xParts := strings.Split(xPkg, "/")
+	yParts := strings.Split(yPkg, "/")
+
+	xRepo := strings.Join(xParts[:3], "/")
+	yRepo := strings.Join(yParts[:3], "/")
+
+	xPkgSuffix := strings.Join(xParts[3:], "/")
+	yPkgSuffix := strings.Join(yParts[3:], "/")
+
+	// Instantiate the type in a separate process as a "Smoke Test"
+	err = instantiate(xRepo, xPkgSuffix, xType)
+	if err != nil {
+		return err
+	}
+	err = instantiate(yRepo, yPkgSuffix, yType)
+	if err != nil {
+		return err
+	}
+
+	// Compare the types by running the template xrt_tmpl.go.tmpl in a separate process
+	// xrt_tmpl.go will return an error if the types are not the same
+	// here we propagate the error to the caller, so as to fail the test.
+	err = serializationDiff(xRepo, xPkgSuffix, xType, yRepo, yPkgSuffix, yType)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func goGet(repo string) error {
+	fmt.Println("Downloading repo:", repo)
+	cmd := exec.Command("go", "get", repo)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	return cmd.Run()
+}
+
+func showKind(pkg, typeName string) error {
+	fmt.Println("Showing kind for:", typeName)
+
+	// Load package
+	cfg := &packages.Config{Mode: packages.NeedName | packages.NeedTypes}
+	pkgs, err := packages.Load(cfg, pkg)
+	if err != nil {
+		return err
+	}
+
+	// Find the type
+	var typ types.Object
+	for _, p := range pkgs {
+		obj := p.Types.Scope().Lookup(typeName)
+		if obj != nil && obj.Exported() {
+			typ = obj
+			break
+		}
+	}
+
+	if typ == nil {
+		return fmt.Errorf("exported type %q not found in package %q", typeName, pkg)
+	}
+
+	// Show the type kind
+	fmt.Printf("Type %q with (kind: %+v)\n\n", typeName, typ.String())
```

Comment:
nit: for future iteration, consider pretty print typ.String().

Reply:
Could you elaborate a bit?

Reply:
Do you mean

```go
	fmt.Printf("Type %q with (kind: %s)\n\n", typeName, typ)
```
?

Reply:
this printed line is hard to read, so I don't find it useful. I was thinking we could either not print this or present it in a better form. 
`(kind: type github.com/algorand/go-algorand/ledger/ledgercore.StateDelta struct{Accts github.com/algorand/go-algorand/ledger/ledgercore.AccountDeltas; KvMods map[string]github.com/algorand/go-algorand/ledger/ledgercore.KvValueDelta; Txids map[github.com/algorand/go-algorand/data/transactions.Txid]github.com/algorand/go-algorand/ledger/ledgercore.IncludedTransactions; ...)`
line breaking at `;` and remove the repo prefix may make it more readable.

Reply:
removed this in [5c1777f](https://github.com/algorand/go-algorand/pull/5304/commits/5c1777f4e639d7d1ed12c8cd66ab65cbb1489d29)